Possible NullPointerException during evaluation

Got an error on one of the nodes:

09:26:30.678 WARN  [tor.default-dispatcher-14] o.e.n.state.UtxoState - Error while applying full block with header 2fbcd449a97d11cab4d2fcf4405481d5257a0110661f5a320f9e0849418bbc7a to UTXOState with root 9d0f3559bdd1fe7dbc36a0a933a8ff9e2e78c6d110cd4f92a5aa036fd8a2c17a18, reason: scorex.core.validation.MalformedModifierError: Scripts of all transaction inputs should pass verification. ee729d6cc0e339b8ab34787b7655fca0e5eb881b84af7cc0c9d2456d4e05f6e5: #1 => Failure(java.lang.NullPointerException) 

For this block: https://explorer.ergoplatform.com/en/blocks/2fbcd449a97d11cab4d2fcf4405481d5257a0110661f5a320f9e0849418bbc7a

Looking for all errors: 

```
20:51:37.856 WARN  [tor.default-dispatcher-12] o.e.m.m.ErgoTransaction - Tx b8153beccbbe74007152de604b2e3d1d69a094677c88d40851832eef8c04b242 verification failed: assertion failed : 
20:51:37.856 WARN  [ctor.default-dispatcher-4] o.e.m.m.ErgoTransaction - Tx 86bb673c35d0cfb6ad54675d67758c4618e1b852f784939bb120c1c8bcbbb8a3 verification failed: null : 
20:51:37.856 WARN  [tor.default-dispatcher-13] o.e.m.m.ErgoTransaction - Tx 41fc07974da4599b2d47cc9c6795dcc77c9eb29fc9a2dbe845cbf7bf25359fae verification failed: null : 
20:51:37.856 WARN  [ctor.default-dispatcher-5] o.e.m.m.ErgoTransaction - Tx 41fc07974da4599b2d47cc9c6795dcc77c9eb29fc9a2dbe845cbf7bf25359fae verification failed: null : 
20:51:37.864 WARN  [tor.default-dispatcher-13] o.e.m.m.ErgoTransaction - Tx 86bb673c35d0cfb6ad54675d67758c4618e1b852f784939bb120c1c8bcbbb8a3 verification failed: null : 
20:51:37.869 WARN  [tor.default-dispatcher-13] o.e.m.m.ErgoTransaction - Tx b8153beccbbe74007152de604b2e3d1d69a094677c88d40851832eef8c04b242 verification failed: null : 
20:51:37.884 WARN  [tor.default-dispatcher-10] o.e.m.m.ErgoTransaction - Tx 0132b36d609bdd742e80d07ac8b178a3cc3287db1a560649424a9bdbc1ca844e verification failed: null : 
20:54:30.414 WARN  [tor.default-dispatcher-11] o.e.m.m.ErgoTransaction - Tx 86bb673c35d0cfb6ad54675d67758c4618e1b852f784939bb120c1c8bcbbb8a3 verification failed: null : 
```


*assertion failed* from above has following details:

```
20:51:37.856 WARN  [tor.default-dispatcher-12] o.e.m.m.ErgoTransaction - Tx b8153beccbbe74007152de604b2e3d1d69a094677c88d40851832eef8c04b242 verification failed: assertion failed : 
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:208)
	at scalan.Base.updateSymbolTable(Base.scala:712)
	at scalan.Base.freshSym(Base.scala:681)
	at scalan.Base$Def.self(Base.scala:164)
	at scalan.Base$Def.self$(Base.scala:163)
	at scalan.Base$CompanionDef.self(Base.scala:212)
	at scalan.Base.$anonfun$reifyObject$1(Base.scala:585)
	at scalan.Base.findOrCreateDefinition(Base.scala:782)
	at scalan.Base.toExp(Base.scala:814)
	at scalan.Base.reifyObject(Base.scala:585)
	at scalan.Library.$anonfun$_specialPredef$1(Library.scala:41)
	at scalan.MutableLazy.value(MutableLazy.scala:13)
	at scalan.Library.specialPredef(Library.scala:42)
	at scalan.Library.specialPredef$(Library.scala:42)
	at sigmastate.eval.RuntimeIRContext.specialPredef(IRContext.scala:145)
	at sigmastate.eval.Evaluation.getDataEnv(Evaluation.scala:333)
	at sigmastate.eval.Evaluation.getDataEnv$(Evaluation.scala:331)
	at sigmastate.eval.RuntimeIRContext.getDataEnv(IRContext.scala:145)
	at sigmastate.eval.IRContext.$anonfun$checkCostWithContext$1(IRContext.scala:129)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.java:23)
	at scala.util.Try$.apply(Try.scala:213)
	at sigmastate.eval.IRContext.checkCostWithContext(IRContext.scala:127)
	at sigmastate.eval.IRContext.checkCostWithContext$(IRContext.scala:126)
	at sigmastate.eval.RuntimeIRContext.checkCostWithContext(IRContext.scala:145)
	at sigmastate.interpreter.PrecompiledScriptReducer.$anonfun$reduce$2(PrecompiledScriptProcessor.scala:84)
	at org.ergoplatform.validation.ValidationRules$.trySoftForkable(ValidationRules.scala:375)
	at sigmastate.interpreter.PrecompiledScriptReducer.reduce(PrecompiledScriptProcessor.scala:81)
	at sigmastate.interpreter.Interpreter.fullReduction(Interpreter.scala:186)
	at sigmastate.interpreter.Interpreter.fullReduction$(Interpreter.scala:174)
	at org.ergoplatform.ErgoLikeInterpreter.fullReduction(ErgoLikeInterpreter.scala:14)
	at sigmastate.interpreter.Interpreter.$anonfun$verify$2(Interpreter.scala:256)
	at scala.util.Try$.apply(Try.scala:213)
	at sigmastate.interpreter.Interpreter.$anonfun$verify$1(Interpreter.scala:230)
	at scalan.util.BenchmarkUtil$.measureTime(BenchmarkUtil.scala:29)
	at sigmastate.interpreter.Interpreter.verify(Interpreter.scala:230)
```


So I guess the issue is in multi-thread access to *_symbolTable* in Base, thus PrecompiledScriptProcessor is not thread-safe probably (in opposite to what its ScalaDoc says)

--- a/core/src/main/scala/scalan/Base.scala
+++ b/core/src/main/scala/scalan/Base.scala
@@ -781,7 +781,6 @@ abstract class Base { scalan: Scalan =>
     if (sym == null) {
       sym = createDefinition(optScope, newSym, d)
     }
-//    assert(te.rhs == d, s"${if (te !) "Found" else "Created"} unequal definition ${te.rhs} with symbol ${te.sym.toStringWithType} for $d")
     sym
   }
 
@@ -792,15 +791,36 @@ abstract class Base { scalan: Scalan =>
     * @return  reference to `d` (which is `s`)
     */
   protected def createDefinition[T](optScope: Nullable[ThunkScope], s: Ref[T], d: Def[T]): Ref[T] = {
-    assert(_symbolTable(d.nodeId).node.nodeId == d.nodeId)
-    assert(s.node eq d, s"Inconsistent Sym -> Def pair $s -> $d")
-    optScope match {
-      case Nullable(scope) =>
-        scope += s
-      case _ =>
-        _globalDefs.put(d, d)
+    try {
+      val nodeId = d.nodeId
+      val tableSym = _symbolTable(nodeId)
+      assert(tableSym.node.nodeId == nodeId)
+      assert(s.node eq d, s"Inconsistent Sym -> Def pair $s -> $d")
+      optScope match {
+        case Nullable(scope) =>
+          scope += s
+        case _ =>
+          _globalDefs.put(d, d)
+      }
+      s
+    } catch { case t: Throwable =>
+      val msg = new mutable.StringBuilder(
+        s"""optScope = $optScope
+        |s = $s
+        |d = $d""".stripMargin)
+      if (d != null) {
+        msg ++= s"\nd.nodeId = ${d.nodeId}"
+        val tableSym = _symbolTable(d.nodeId)
+        msg ++= s"\n_symbolTable(d.nodeId) = $tableSym"
+        if (tableSym != null) {
+          msg ++= s"\ntableSym.node = ${tableSym.node}"
+          if (tableSym.node != null) {
+            msg ++= s"\ntableSym.node.nodeId = ${tableSym.node.nodeId}"
+          }
+        }
+      }
+      throw new RuntimeException(msg.result(), t)
     }
-    s
   }
 
   /**

--- a/core/src/main/scala/scalan/Base.scala
+++ b/core/src/main/scala/scalan/Base.scala
@@ -709,7 +709,14 @@ abstract class Base { scalan: Scalan =>
       cfor(0)(_ < delta, _ + 1) { _ => tab.append(null) }
       val sym = if (s == null) new SingleRef(d) else s
       tab += sym
-      assert(tab.length == id + 1)
+      assert(tab.length == id + 1,
+        s"""tab.length == id + 1:
+          |tab.length = ${tab.length}
+          |id = $id
+          |s = $s
+          |d = $d
+          |sym = $sym
+          |""".stripMargin)
       sym
     }
   }

--- a/sigmastate/src/main/scala/sigmastate/eval/IRContext.scala
+++ b/sigmastate/src/main/scala/sigmastate/eval/IRContext.scala
@@ -1,18 +1,22 @@
 package sigmastate.eval
 
-import java.lang.{Math => JMath}
 import sigmastate.SType
-import sigmastate.Values.{Value, SValue}
+import sigmastate.Values.{SValue, Value}
 import sigmastate.interpreter.Interpreter.ScriptEnv
 import sigmastate.lang.TransformingSigmaBuilder
 import sigmastate.lang.exceptions.CostLimitException
+import sigmastate.utils.Helpers
 import sigmastate.utxo.CostTable
 
+import java.util.concurrent.locks.ReentrantLock
 import scala.util.Try
 
 trait IRContext extends Evaluation with TreeBuilding {
 
   override val builder = TransformingSigmaBuilder
+
+  /** Can be used to synchronize access to this IR object from multiple threads. */
+  val lock = new ReentrantLock()
 
   /** Pass configuration which is used to turn-off constant propagation.
     * @see `beginPass(noCostPropagationPass)`  */
@@ -125,9 +129,12 @@ trait IRContext extends Evaluation with TreeBuilding {
     */
   def checkCostWithContext(ctx: SContext,
                 costF: Ref[((Context, (Int, Size[Context]))) => Int], maxCost: Long, initCost: Long): Try[Int] = Try {
-    val costFun = compile[(SContext, (Int, SSize[SContext])), Int, (Context, (Int, Size[Context])), Int](
-                    getDataEnv, costF, Some(maxCost))
-    val (estimatedCost, accCost) = costFun((ctx, (0, Sized.sizeOf(ctx))))
+
+    val (estimatedCost, accCost) = Helpers.withReentrantLock(lock) { // protect mutable access to this IR
+      val costFun = compile[(SContext, (Int, SSize[SContext])), Int, (Context, (Int, Size[Context])), Int](
+        getDataEnv, costF, Some(maxCost))
+      costFun((ctx, (0, Sized.sizeOf(ctx))))
+    }
 
     if (debugModeSanityChecks) {
       if (estimatedCost != accCost)

--- a/sigmastate/src/main/scala/sigmastate/interpreter/PrecompiledScriptProcessor.scala
+++ b/sigmastate/src/main/scala/sigmastate/interpreter/PrecompiledScriptProcessor.scala
@@ -2,17 +2,17 @@ package sigmastate.interpreter
 
 import java.util.concurrent.ExecutionException
 import java.util.concurrent.atomic.AtomicInteger
-
-import com.google.common.cache.{CacheBuilder, RemovalNotification, RemovalListener, LoadingCache, CacheLoader, CacheStats}
+import com.google.common.cache.{CacheBuilder, CacheLoader, CacheStats, LoadingCache, RemovalListener, RemovalNotification}
 import org.ergoplatform.settings.ErgoAlgos
 import org.ergoplatform.validation.SigmaValidationSettings
-import org.ergoplatform.validation.ValidationRules.{CheckCostFunc, CheckCalcFunc, trySoftForkable}
+import org.ergoplatform.validation.ValidationRules.{CheckCalcFunc, CheckCostFunc, trySoftForkable}
 import scalan.{AVHashMap, Nullable}
 import sigmastate.Values
 import sigmastate.Values.ErgoTree
-import sigmastate.eval.{RuntimeIRContext, IRContext}
+import sigmastate.eval.{IRContext, RuntimeIRContext}
 import sigmastate.interpreter.Interpreter.{ReductionResult, WhenSoftForkReductionResult}
 import sigmastate.serialization.ErgoTreeSerializer
+import sigmastate.utils.Helpers
 import sigmastate.utils.Helpers._
 import spire.syntax.all.cfor
 
@@ -84,9 +84,11 @@ case class PrecompiledScriptReducer(scriptBytes: Seq[Byte])(implicit val IR: IRC
       val estimatedCost = IR.checkCostWithContext(costingCtx, costF, maxCost, initCost).getOrThrow
 
       // check calc
-      val calcF = costingRes.calcF
       val calcCtx = context.toSigmaContext(isCost = false)
-      val res = Interpreter.calcResult(IR)(calcCtx, calcF)
+      val res = Helpers.withReentrantLock(IR.lock) { // protecting mutable access to IR instance
+        val calcF = costingRes.calcF
+        Interpreter.calcResult(IR)(calcCtx, calcF)
+      }
       ReductionResult(SigmaDsl.toSigmaBoolean(res), estimatedCost)
     }
   }

--- a/sigmastate/src/main/scala/sigmastate/utils/Helpers.scala
+++ b/sigmastate/src/main/scala/sigmastate/utils/Helpers.scala
@@ -1,6 +1,5 @@
 package sigmastate.utils
 
-import java.util
 import io.circe.Decoder
 import org.ergoplatform.settings.ErgoAlgos
 import sigmastate.eval.{Colls, SigmaDsl}
@@ -8,7 +7,8 @@ import sigmastate.interpreter.CryptoConstants.EcPointType
 import special.collection.Coll
 import special.sigma.GroupElement
 
-import java.util.concurrent.locks.{Lock, ReentrantLock}
+import java.util
+import java.util.concurrent.locks.Lock
 import scala.reflect.ClassTag
 import scala.util.{Either, Failure, Right, Success, Try}
 

--- a/sigmastate/src/main/scala/sigmastate/utils/Helpers.scala
+++ b/sigmastate/src/main/scala/sigmastate/utils/Helpers.scala
@@ -1,7 +1,6 @@
 package sigmastate.utils
 
 import java.util
-
 import io.circe.Decoder
 import org.ergoplatform.settings.ErgoAlgos
 import sigmastate.eval.{Colls, SigmaDsl}
@@ -9,8 +8,9 @@ import sigmastate.interpreter.CryptoConstants.EcPointType
 import special.collection.Coll
 import special.sigma.GroupElement
 
+import java.util.concurrent.locks.{Lock, ReentrantLock}
 import scala.reflect.ClassTag
-import scala.util.{Failure, Try, Either, Success, Right}
+import scala.util.{Either, Failure, Right, Success, Try}
 
 object Helpers {
   def xor(ba1: Array[Byte], ba2: Array[Byte]): Array[Byte] = ba1.zip(ba2).map(t => (t._1 ^ t._2).toByte)
@@ -157,6 +157,27 @@ object Helpers {
   def decodeBytes(base16String: String): Coll[Byte] = {
     val bytes = ErgoAlgos.decodeUnsafe(base16String)
     Colls.fromArray(bytes)
+  }
+
+  /**
+    * Executes the given block with a reentrant mutual exclusion Lock with the same basic
+    * behavior and semantics as the implicit monitor lock accessed using synchronized
+    * methods and statements in Java.
+    *
+    * Note, using this method has an advantage of having this method in a stack trace in case of
+    * an exception in the block.
+    * @param l lock object which should be acquired by the current thread before block can start executing
+    * @param block block of code which will be executed retaining the lock
+    * @return the value produced by the block
+    */
+  def withReentrantLock[A](l: Lock)(block: => A): A = {
+    l.lock()
+    val res = try
+      block
+    finally {
+      l.unlock()
+    }
+    res
   }
 }
 

